Fix issue with variables not substituted in `uri`-referenced manifests (#5451)

rm3l opened the pull request:
**What type of PR is this:**
/kind bug

**What does this PR do / why we need it:**

This fixes the issue reported in #5451, where running `odo deploy` (actually any commands calling `libdevfile.GetK8sComponentAsUnstructured`) using a Devfile containing `uri`-referenced manifests containing variables would  not succeed, either with a YAML validation error message similar to the line below or with an error (e.g., unresolvable image name) in the manifest once deployed in the cluster:

```
error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{"appname":interface {}(nil)}
```

**Which issue(s) this PR fixes:**
Fixes #5451 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
Refer to the reproduction steps depicted in the original issue for testing this: #5451 

From the code perspective, as discussed during one of the Sprint Plannings, this PR is enriching the `libdevfile` package with a new `GetK8sManifestWithVariablesSubstituted` function that is responsible for returning the resource manifest from the Kubernetes/OpenShift Devfile component with all variables substituted, rather than having `libdevfile.GetK8sComponentAsUnstructured` (and few other places in the codebase) fetch the raw manifest from the URI and handle it as is.
Also, the implementation aims at leaving room for supporting `OpenShift` components in a near future (scope of #5704).

~Since this seems a much more realistic scenario of how outer-loop manifests are referenced in Devfiles, this PR also switches to `uri` references in all test Devfiles. This is the purpose of df1f19ed959b8c82aa17ac3cf12c343305789149
Note that tests for `odo deploy` still test both `inlined` (as before) and `uri` manifests (scope of this issue).~ [Will be covered in a separate PR]

feloy commented on pkg/libdevfile/libdevfile.go:
```diff
@@ -193,3 +199,83 @@ func GetEndpointsFromDevfile(devfileObj parser.DevfileObj, ignoreExposures []v1a
+// No matter how the component is defined, it returns
+// the content with all variables substituted using the global variables map defined in `devfileObj`.
+// An error is returned if the content references an invalid variable key not defined in the Devfile object.
+func GetComponentResourceManifestContentWithVariablesResolved(devfileObj parser.DevfileObj, devfileCmp interface{},
```
Now we get the entire devfileObj as a parameter, you could pass the name of the component instead of its content

rm3l replied:
Is there a simple way to retrieve a Component by its name in `devfileObj`? It is not clear in the docs, but are we sure component names are unique in the Devfile? (Maybe by both name and type?)

Or maybe I could avoid passing the `devfileObj`, and use the variables map as an argument instead. This is what I actually used `devfileObj` for, to substitute the variables using: `devfileObj.Data.GetDevfileWorkspaceSpec().Variables`

WDYT?

rm3l replied:
Ah, I found a `Filter` field in `DevfileOptions` that could allow to retrieve Components by their name. But still not sure about the unicity.

rm3l replied:
Just found our `ComponentsWithSameNameError` error type we already have in `libdevfile`. This seems to answer my previous question :)
So I will make the changes you suggested, using the component name instead.

feloy commented on tests/integration/devfile/cmd_devfile_build_images_test.go:
```diff
@@ -79,6 +79,9 @@ var _ = Describe("odo devfile build-images command tests", func() {
 
 		BeforeEach(func() {
 			helper.CopyExample(filepath.Join("source", "nodejs"), commonVar.Context)
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project", "kubernetes",
```
Should odo copy these dependencies when running `odo init --devfile-path`? @kadel WDYT?

rm3l replied:
+1. Indeed, that would be a nice feature, I think.

rm3l replied:
For reference, #5713 has been created for this point.

feloy reviewed (commented):
You have moved devfile from using `Inlined` to using `URI` as discussed during issue preparation. Are there some remaining devfile using `Inlined`? I would not like we make regressions on this now :(

rm3l commented on pkg/libdevfile/errors.go:
```diff
@@ -64,14 +64,18 @@ func (e ComponentNotExistError) Error() string {
 
 func (e ComponentsWithSameNameError) Error() string {
-	return "more than one component with the same name, should not happen"
+	return fmt.Sprintf("more than one component with the same name (%q), should not happen", e.name)
```
Good catch. I'll change it - thanks.

rm3l commented:
> You have moved devfile from using `Inlined` to using `URI` as discussed during issue preparation. Are there some remaining devfile using `Inlined`? I would not like we make regressions on this now :(

I identified the following places:
- [`devfile-deploy-with-k8s-inlined.yaml`](https://github.com/redhat-developer/odo/pull/5711/files#diff-1e678e962ad0dcb3c1e54400840d6e085daba732288da5692d2464adc40d0a89) (actually, it is just the old `devfile-deploy.yaml` file that I renamed). Renaming it was intentional to make sure 'Inlined' manifests could also be tested with `odo deploy` as before.
- [`devfile-deploy-with-multiple-resources.yaml`](https://github.com/redhat-developer/odo/pull/5711/files#diff-81ecd14331e084235ba14a08e423561e6aedded41486f39f137385f065a92405) : this one can't use URI references at this time because of #5637. I added `TODO` comments in this file.

I was also hesitant on changing this in this PR. That's why the commit is isolated if we want to revert or do this change in a separate issue/PR.

Ideally, I think we should come up with a way to easily reuse the same tests against both Inlined and URI-referenced manifests. This is what I tried to accomplish with this loop in [`cmd_devfile_deploy_test.go`](https://github.com/redhat-developer/odo/pull/5711/files#diff-ffa8db34cac925d833d18af25ef37c7e7466ea55253fd810215ce600c3d9c01b).

rm3l commented:
> > You have moved devfile from using `Inlined` to using `URI` as discussed during issue preparation. Are there some remaining devfile using `Inlined`? I would not like we make regressions on this now :(
> 
> I identified the following places:
> 
> * [`devfile-deploy-with-k8s-inlined.yaml`](https://github.com/redhat-developer/odo/pull/5711/files#diff-1e678e962ad0dcb3c1e54400840d6e085daba732288da5692d2464adc40d0a89) (actually, it is just the old `devfile-deploy.yaml` file that I renamed). Renaming it was intentional to make sure 'Inlined' manifests could also be tested with `odo deploy` as before.
> * [`devfile-deploy-with-multiple-resources.yaml`](https://github.com/redhat-developer/odo/pull/5711/files#diff-81ecd14331e084235ba14a08e423561e6aedded41486f39f137385f065a92405) : this one can't use URI references at this time because of [odo delete fails for kubernetes components #5637](https://github.com/redhat-developer/odo/issues/5637). I added `TODO` comments in this file.
> 
> I was also hesitant on changing this in this PR. That's why the commit is isolated if we want to revert or do this change in a separate issue/PR.
> 
> Ideally, I think we should come up with a way to easily reuse the same tests against both Inlined and URI-referenced manifests. This is what I tried to accomplish with this loop in [`cmd_devfile_deploy_test.go`](https://github.com/redhat-developer/odo/pull/5711/files#diff-ffa8db34cac925d833d18af25ef37c7e7466ea55253fd810215ce600c3d9c01b).

To be safe, I just reverted that commit from this PR, but saved it to a separate branch, and will try to include it in a dedicated PR later on.

feloy commented:
/approve

rm3l commented:
Just rebased onto `main` to benefit from #5712, which fixed the tests.

feloy commented:
/lgtm

rm3l commented:
/override ci/prow/v4.10-integration-e2e

Looks like a problem not related to this PR, but related to the deprovisioning job that tears down the cluster on Prow: 

>   * could not run steps: step integration-e2e failed: "integration-e2e" post steps failed: "integration-e2e" pod "integration-e2e-ipi-deprovision-deprovision" failed: the pod ci-op-kzni8cn5/integration-e2e-ipi-deprovision-deprovision was deleted without completing after 4m50s (failed containers: )

All tests passed on IBM Cloud by the way.
